jep/0000/README.adoc

```diff
@@ -0,0 +1,264 @@
+The Bill of Materials (**BoM**) aims to set a common language that can be reused across them.
+
+The BoM is trying to be a cornerstone to enable these other goals:
+
+* Ensure value gets delivered while quality is continuously improved
+* Do not go down the rabbit hole of spending weeks without delivering value
+* Increase quality of non-LTS builds
+* Provide an answer to “Is this code ready to be delivered to Jenkins Essentials?”
+* Avoid complex matrix testing that is the objective of Essentials tracing and monitoring
+* Feedback delivered to PR author as soon as possible
+* Automated releases - at least from the perspective of the engineer
+* Reduce code review overheads so that maintainers can process more.
+AKA “I am a plugin maintainer, not a FindBugs robot. Automate checks for me”
+
+
+== Specification
+
+=== What exists today
+
+* https://github.com/jenkinsci/acceptance-test-harness[Jenkins acceptance test harness]
+* https://github.com/jenkinsci/blueocean-acceptance-test[Blue Ocean acceptance test]
+* https://github.com/jenkinsci/plugin-compat-tester[Jenkins plugin compat tester]
+* https://github.com/jenkins-infra/evergreen[Essentials evergreen draft]
+* https://github.com/jenkinsci/jenkins-war-packager[Jenkins WAR packager] recently started
+
+=== What are we missing
+
+* Builds/Tests from master, needed for Essentials
+* Builds/Tests from core, libraries and plugins PRs
+
+These cases are addressed by a Custom WAR Packager prototype for custom WAR Building and testing use-cases.
+That format is narrow focused, and it does not address use-cases like support of custom environments.
+We want to unify the description format across projects within Jenkins (e.g. Jenkins Essentials, Jenkins X, Docker Packaging, etc.).
+
+=== Bill of Materials (BoM)
+
+The BoM defines what commits of core, bundled libraries (stapler, remoting)  and plugins are delivered and tested together.
+They key part is the usage of commit ids and not releases.
+The usage of the maven release plugin does not fit this requirement.
+
+The BoM needs to be trackable in Git.
+
+The BoM needs to have the concept of environments for Essentials. Ie. install different plugins in AWS vs Azure vs...
+
+A requirement for Essentials is that parts of the BoM need to be downloadable.
+Essentials needs to enable selective downloads of core, bundled libraries and plugins for incremental upgrades.
+Pushing into maven repo is fine, although indexing time is big.
+Will need to mirror into Azure CDN at some point.
+
+The BoM needs to be translated to take advantage of existing infrastructure.
+
+* Essentials needs to generate a diff for upgrades and download the bits
+* Existing ATH uses a war binary that can be built from BoM
+
+Packages should be trackable back to a commit id, but during development the BoM should point to branches or tags, so we would have a BoM as input and a effective BoM as output with all the git refs resolved to specific commit ids.
```

Comment:
> more useful is for it to refer to a `SNAPSHOT` version in that case

_Most_ useful would be for it to refer to a local file path, triggering a build from sources. I think the WAR packager has an option like this.

Reply:
I do like the idea of freeing developers from tedious tasks, but as much as I like that idea, I am not sure we here have a clear audit trail/traceability of what _versions_ exactly go to production. And I do deeply dislike the idea of "version range" or something similar. And saying one uses a given branch is similar to this: you don't really know what you're using, and it can change somewhere.

I prefer the _GitOps_ approach described in JEP-305 where things are explicitly updated using a PR for instance. I think we must design things for end-user consumption, _then_ we can think about how to create tooling to make things easier to bump/update locally for quick development turnaround.

_Maybe_ here what you have in mind is not contradictory to this, but I'm just voicing my concern in case.